### PR TITLE
Balder/reduce get node calls

### PR DESCRIPTION
**Pull request by baldersheim**
@vekterli PR

### CODE REVIEW

File: storage/src/vespa/storage/distributor/operations/idealstate/setbucketstateoperation.cpp [hunk 1]
```diff
@@ -34,8 +35,8 @@ SetBucketStateOperation::enqueueSetBucketStateCommand(uint16_t node, bool active
 bool
 SetBucketStateOperation::shouldBeActive(uint16_t node) const
 {
-    for (uint32_t i=0, n=_wantedActiveNodes.size(); i<n; ++i) {
-        if (_wantedActiveNodes[i] == node) {
+    for (unsigned short _wantedActiveNode : _wantedActiveNodes) {
```
Comment (vekterli) [inline storage/src/vespa/storage/distributor/operations/idealstate/setbucketstateoperation.cpp:38]: Consider `uint16_t` and avoiding leading underscore for local variable

File: storage/src/vespa/storage/distributor/operations/idealstate/setbucketstateoperation.cpp [hunk 2]
```diff
@@ -44,8 +45,8 @@ SetBucketStateOperation::shouldBeActive(uint16_t node) const
 
 void
 SetBucketStateOperation::activateNode(DistributorStripeMessageSender& sender) {
-    for (uint32_t i=0; i<_wantedActiveNodes.size(); ++i) {
-        enqueueSetBucketStateCommand(_wantedActiveNodes[i], true);
+    for (unsigned short _wantedActiveNode : _wantedActiveNodes) {
```
Comment (vekterli) [inline storage/src/vespa/storage/distributor/operations/idealstate/setbucketstateoperation.cpp:48]: Consider `uint16_t` and avoiding leading underscore for local variable

File: storage/src/vespa/storage/distributor/operations/idealstate/setbucketstateoperation.cpp [hunk 3]
```diff
@@ -55,9 +56,9 @@ SetBucketStateOperation::activateNode(DistributorStripeMessageSender& sender) {
 void
 SetBucketStateOperation::deactivateNodes(DistributorStripeMessageSender& sender) {
     const std::vector<uint16_t>& nodes(getNodes());
-    for (size_t i = 0; i < nodes.size(); ++i) {
-        if (!shouldBeActive(nodes[i])) {
-            enqueueSetBucketStateCommand(nodes[i], false);
+    for (unsigned short node : nodes) {
```
Comment (vekterli) [inline storage/src/vespa/storage/distributor/operations/idealstate/setbucketstateoperation.cpp:59]: Consider `uint16_t`

**Review by vekterli — DISMISSED** [review]
:+1: I am a bit concerned about how closely this couples the active bucket result and the original DB entry it was generated from (due to the former "pointing into" the latter), but hopefully the two should never be used independently in practice.

File: storage/src/vespa/storage/distributor/operations/idealstate/setbucketstateoperation.cpp [hunk 4]
```diff
@@ -10,6 +10,10 @@ LOG_SETUP(".distributor.operation.idealstate.setactive");
 
 namespace storage::distributor {
 
+namespace {
+
+}
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/vespa/storage/distributor/operations/idealstate/setbucketstateoperation.cpp]: Fixed

File: storage/src/vespa/storage/distributor/operations/idealstate/setbucketstateoperation.cpp (hunk 2 above)
Comment (baldersheim) [reply storage/src/vespa/storage/distributor/operations/idealstate/setbucketstateoperation.cpp]: hmm, I guess I have to double check when accepting CLions rewrites.
Fixed.

File: storage/src/vespa/storage/distributor/operations/idealstate/setbucketstateoperation.cpp (hunk 1 above)
Comment (baldersheim) [reply storage/src/vespa/storage/distributor/operations/idealstate/setbucketstateoperation.cpp]: Fixed

File: storage/src/vespa/storage/distributor/operations/idealstate/removebucketoperation.cpp [hunk 5]
```diff
@@ -20,8 +20,7 @@ RemoveBucketOperation::onStartInternal(DistributorStripeMessageSender& sender)
 
     BucketDatabase::Entry entry = _bucketSpace->getBucketDatabase().get(getBucketId());
 
-    for (uint32_t i = 0; i < getNodes().size(); ++i) {
-        uint16_t node = getNodes()[i];
+    for (unsigned short node : getNodes()) {
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/vespa/storage/distributor/operations/idealstate/removebucketoperation.cpp:23]: Fixed

File: storage/src/vespa/storage/bucketdb/bucketinfo.hpp [hunk 6]
```diff
@@ -144,7 +144,18 @@ BucketInfoBase<NodeSeq>::getNode(uint16_t node) const noexcept {
...
-    return 0;
+    return nullptr;
+}
+
+template <typename NodeSeq>
+uint16_t
+BucketInfoBase<NodeSeq>::getRef(uint16_t node) const noexcept {
+    for (const auto& n : _nodes) {
+        if (n.getNode() == node) {
+            return &n - &_nodes[0];
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/vespa/storage/bucketdb/bucketinfo.hpp]: Fixed

File: storage/src/vespa/storage/bucketdb/bucketinfo.h [hunk 7]
```diff
@@ -86,6 +86,7 @@ class BucketInfoBase
      * Returns the bucket copy struct for the given node, null if nonexisting
      */
     const BucketCopy* getNode(uint16_t node) const noexcept;
+    uint16_t getRef(uint16_t node) const noexcept;
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/vespa/storage/bucketdb/bucketinfo.h]: Fixed

File: storage/src/vespa/storage/distributor/activecopy.h [hunk 8]
```diff
@@ -19,13 +19,15 @@ class ActiveCopy {
         : _nodeIndex(Index::invalid()),
           _ideal(Index::invalid()),
           _doc_count(0),
+          _entryIndex(Index::invalid()),
           _ready(false),
           _active(false)
     { }
-    ActiveCopy(uint16_t node, const BucketCopy & copy, uint16_t ideal) noexcept
+    ActiveCopy(uint16_t node, const BucketCopy & copy, uint16_t ideal, uint32_t entryIndex_in) noexcept
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/vespa/storage/distributor/activecopy.h]: Fixed

File: storage/src/tests/distributor/top_level_bucket_db_updater_test.cpp [hunk 9]
```diff
@@ -1427,99 +1404,62 @@ TopLevelBucketDBUpdaterTest::get_sent_nodes_distribution_changed(const std::stri
...
 
     EXPECT_EQ(get_node_list({2, 3, 4, 5}),
-              get_sent_nodes("distributor:1 storage:2",
-                             "distributor:1 storage:6"));
+              get_sent_nodes("distributor:1 storage:2", "distributor:1 storage:6"));
 
     EXPECT_EQ(get_node_list({0, 1, 2}),
-              get_sent_nodes("distributor:4 storage:3",
-                             "distributor:3 storage:3"));
+              get_sent_nodes("distributor:4 storage:3", "distributor:3 storage:3"));
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/tests/distributor/top_level_bucket_db_updater_test.cpp]: PTAL

**Review by vekterli — APPROVED** [review]
:+1:

**Pull request merged**